### ISSUE TITLE
tf.distribute.MultiWorkerStrategy - Don't mirror models on workers.

### ISSUE THREAD

**Issue by stefan-falk** [tensorflow/tensorflow#52197]
I am exploring multi-worker training with Tensorflow and I'm at the point where I would need a `tf.distribute.MultiWorkerStrategy`. Compare `tf.distribute.Strategy` with the `tf.distribute.MirroredStrategy`.

I do not want the model to get mirrored on each GPU of each worker, instead I want the model to get mirrored on each worker.

Is there a way I can to that? Would be possible to implement tf.distribute.MultiWorkerStrategy` by myself (without too much effort)?

**Comment by tilakrayal** [comment]
@stefan-falk ,
Can you please elaborate about your Feature. Also, please specify the Use Cases for this feature. Thanks!

**Comment by stefan-falk** [comment]
@tilakrayal Of course. 

After looking into this though I came to the conclusion that what I want is probably not possible (yet). The reason for this is the fact that all gradients are computed on one device at the moment and it does not seem to be a trivial task to distribute these computations over multiple devices.

But let me explain where I am coming from. Maybe there's a simpler solution to my problem already available.

----

Now, usually when we use the classic `MirroredStrategy` we'd have the following (e.g. for 4 GPUs on one machine):

```none
device | stores
---------------
gpu:0  | model, gradients
gpu:1  | model
gpu:2  | model 
gpu:3  | model  
```

In my case this is a problem because I'll run into OOM exceptions on `/gpu:0` rather quickly once the model is large enough. 

### Scenario 1

One solution here would be to remove the `model` from `/gpu:0` s.t. we can increase the model or batch-size:

```none
device | stores
---------------
gpu:0  | gradients
gpu:1  | model
gpu:2  | model 
gpu:3  | model  
```

I am not sure how I could achieve such a setup locally / on a single worker.

### Scenario 2

However, even if scenario 1 was possible, I'd already loose one model instance and the model is still very limited by the VRAM of just one device. This is where I realized that I could distribute the model across multiple GPUs:

```none
device | stores
---------------
gpu:0  | encoder_layers[0:10]  & gradients
gpu:1  | encoder_layers[10:20] & gradients
gpu:2  | decoder_layers[0:10]  & gradients
gpu:3  | decoder_layers[10:20] & gradients
```

If this was possible, it would be useful to have the suggested `tf.distribute.MultiWorkerStrategy` s.t. this setup gets exactly replicated on each worker/machine.

However, the problem here are the gradients. It appears to be (almost) impossible to distribute the computations for the gradients across multiple GPUs at this point. For this reason, scenario 2 is probably off the table I guess?

**Comment by sachinprasadhs** [comment]
Have you checked [ParameterServerStrategy](https://www.tensorflow.org/guide/distributed_training#parameterserverstrategy) which may help your case, more details can be found in [this](https://www.tensorflow.org/api_docs/python/tf/distribute/experimental/coordinator/ClusterCoordinator) document with the example. Let us know if this helps you. Thank you!

**Comment by rchao** [comment]
Hello @stefan-falk thanks for the question! I'd suggest visiting https://www.tensorflow.org/guide/distributed_training and see if either [`ParameterServerStrategy`](https://www.tensorflow.org/tutorials/distribute/parameter_server_training) or [`MultiWorkerMirroredStrategy`](https://www.tensorflow.org/tutorials/distribute/multi_worker_with_keras) suits your need. I would give them a try.

**Comment by rchao** [comment]
Sounds good. Thanks!

**Comment by sachinprasadhs** [comment]
@stefan-falk , Have found a solution with the above mentioned workaround, if so could you please close the issue. Thanks!
